### ISSUE TITLE
P4 DPDK on horizon; BMV2

### ISSUE THREAD

**Issue by KrisNey-MSFT** [sonic-net/DASH#408]
DPDK BE test/experiment
The engineers who work on p4c-dpdk back end have been notified (and provided some of the details on the reasons for failure to load that I included in the issues).  These issues are not their highest priority tasks, so not clear when they might be able to address them.

Compiles fine, needs to load into the software switch.
	
Just experimental, open source P4 BE DPDK code that produces the spec file (possible bug), but still low priority - feel free to check them out!  How does P4C work???
	
Intel Eng not opposed to pursuing resolution of issues; need to check w/Reshma to see if issues will be fixed within a timeframe.  

https://github.com/p4lang/p4c/tree/main/backends/dpdk
https://github.com/p4lang/p4c/issues?q=is%3Aissue+is%3Aopen+label%3Adash-blocker

Usha Gupta is coordinating (GitHub ID)
Ping Github id @usha1830 if you are interested in discussing those issues in more detail.
This is Usha Gupta, Intel engineer who has developed most of the p4c-dpdk back end code.
You can also ping me at @jafingerhut if I might be of help, but I'm more familiar with the symptoms than the p4c-dpdk back end C++ code.

**Comment by usha1830** [comment]
@KrisNey-MSFT @reshmaintel 
The issues mentioned in https://github.com/p4lang/p4c/issues?q=is%3Aissue+is%3Aopen+label%3Adash-blocker are fixed and merged in p4c main branch. 
Please let me know if there are any further issues.

**Comment by jafingerhut** [comment]
I have tried it with the version of DASH at the time I filed that issue in late March/early April, and there are no errors compiling that version, and no errors loading it into the P4-DPDK software switch.

I then tested compiling the latest version of DASH P4 code.  It compiles with warnings (but no errors -- mostly warnings about unused parameters/fields/extern instances), but it gives errors when attempting to load into the P4-DPDK software switch.  I have not root-caused it, but will pass on how to reproduce it to some Intel engineers who know how to do that more quickly than I do.

**Comment by jafingerhut** [comment]
I have created a new issue for p4c-dpdk to track this latest discovered "dash blocker": https://github.com/p4lang/p4c/issues/4156

I plan to use the "dash blocker" tag for any such issues, to make them easier to find: https://github.com/p4lang/p4c/issues?q=is%3Aissue+is%3Aopen+label%3Adash-blocker

**Comment by KrisNey-MSFT** [comment]
@jafingerhut - should we close this one, and continue work with 4156?  Or keep this one open for tracking? TY, KrisNey-MSFT.

**Comment by jafingerhut** [comment]
If you want a DASH-specific place to track what is missing for DASH, it seems best to continue to have an issue in this repo for that purpose.  It seems reasonable to continue to use this issue until P4-DPDK can pass a first packet through for whatever is at the time the current version of the DASH P4 reference code.

**Comment by KrisNey-MSFT** [comment]
Andy - issue with P4-DPDK backend. The code at [https://github.com/sonic-net/DASH/blob/main/dash-pipeline/bmv2/dash_service_tunnel.p4#L7-L24] which performs 128-bit masking on IPv6 addresses exposes a deficiency in p4-dpdk backend, it can't perform 128-bit math. 
Action - @jafingerhut will pursue it with the p4-dpdk team. He may also propose a workaround/compromise and file as an issue or PR.

**Comment by KrisNey-MSFT** [comment]
Per 9/27/2023 Community Call, Andy is going to lobby for these enhancements in P4 group.  
Can check back another week or so.

**Comment by KrisNey-MSFT** [comment]
Same status, check next week

**Comment by KrisNey-MSFT** [comment]
Ping sent to Intel team

**Comment by KrisNey-MSFT** [comment]
@usha1830 fixed 1 issue, working on the 2nd.  
Defined a header w/bit fields not multiples of 8, the P4 DPDK implementation was swapping these in the wire order of bits.  Not DASH specific, but was likely fixed.

Current DASH P4 opers on 128bit - people at Intel know about this.  

@ashutosh-agrawal discussing w/P4 DPDK C target team.  Right now only support 64bit operation.  Not the compiler, it's the target.  2 different components need to be fixed.  DASH just needs bit-wise... and an or, sub and add, shift...etc...

**Comment by KrisNey-MSFT** [comment]
Waiting for Intel staff for answers to the above.

**Comment by KrisNey-MSFT** [comment]
Hi @ashutosh-agrawal  and @usha1830  - I didn't see you in the call today.  We were curious whether there is an update to this Issue?  TY, Kristina

**Comment by KrisNey-MSFT** [comment]
Compiler and soft switch work.  PRs pending...

**Comment by KrisNey-MSFT** [comment]
P4 DPDK will be faster, yet heavy lifting required to design

**Comment by jafingerhut** [comment]
An update from some experiments run in the last few days:  I found additional issues that seem like they would block DASH using P4-DPDK.  As far as I can tell, these are not issues with the p4c-dpdk P4 compiler, but with the P4Runtime API server implementation for P4-DPDK.  I have filed these two public issues asking questions about the difficulties I have found:

+ https://github.com/ipdk-io/networking-recipe/issues/423
+ https://github.com/ipdk-io/networking-recipe/issues/424

**Comment by chrispsommers** [comment]
Hi Andy, thanks for exploring this and keeping us up to date!

**Comment by KrisNey-MSFT** [comment]
Looking to find resource at Intel, Usha is no longer with the team.  We cannot run a subset of tests with either bmv2 or P4 DPDK.  Priority = Medium.

**Comment by KrisNey-MSFT** [comment]
Sosutha (possible resource) from Intel

**Comment by KrisNey-MSFT** [comment]
hi @Sosutha - although we didn't have time during the call today...would you happen to have an update for this one?  We can discuss tomorrow in Behavioral Model call :)

**Comment by Sosutha** [comment]
Implementation for 128 bit support for bitwise instructions is started now

**Comment by KrisNey-MSFT** [comment]
Thank you @Sosutha !  @r12f and @jafingerhut for visiblity...

**Comment by KrisNey-MSFT** [comment]
PR perhaps open next week in open source.

**Comment by Sosutha** [comment]
PR is in progress. Testcase addition in progress. It will be raised in open source this week

**Comment by KrisNey-MSFT** [comment]
Could try this on your fork @Sosutha ?  Is this published to docker pub?  Or wait for the PR to be merged in order to use the docker file?

**Comment by Sosutha** [comment]
@KrisNey-MSFT, It is not clear to me what you are asking me to try. PR for 128bit bitwise operations in dpdk is here. https://github.com/p4lang/p4c/pull/4952

**Comment by KrisNey-MSFT** [comment]
[Link to Intel PR](https://github.com/p4lang/p4c/pull/4952)

**Comment by chrispsommers** [comment]
> Could try this on your fork @Sosutha ? Is this published to docker pub? Or wait for the PR to be merged in order to use the docker file?

Hi @Sosutha, this request originally came from me. What I meant was, could you try out this DPDK version with the DASH build pipeline to make sure it compiles P4 with DPDK, including the 128-bit operations? The purpose would be to make sure the DPDK enhancements actually solved the problems we found in DASH, ideally before you committed to p4c.

It would require forking DASH and making a branch with your updated p4c as a docker image, then building DASH, at least the [p4-dpdk-pna](https://github.com/sonic-net/DASH/blob/e16ceb457c9453557f584468c6ed789a8b5a1bfa/dash-pipeline/Makefile#L141C1-L141C13) make target. You'd also need to remove the flag `-DDISABLE_128BIT_ARITHMETIC` from https://github.com/sonic-net/DASH/blob/e16ceb457c9453557f584468c6ed789a8b5a1bfa/dash-pipeline/Makefile#L154 in the Makefile to expose the 128 bit math and verify your fix.

For reference, here is how @jafingerhut updated the DASH repo with a new version of p4c-dpdk: https://github.com/sonic-net/DASH/pull/512

This might be a chicken and egg problem because the p4c docker might not get published until your PR is merged into p4c. The alternative is to wait for your DPDK PR to merge, *then* do a fork of DASH and make a branch to update the p4c-dpdk docker version similar to https://github.com/sonic-net/DASH/pull/512. In any case, once your p4c-dpdk fix is merged, we'll need a volunteer to file a PR for DASH with the changes I described above (Makefile and Dockerfile).

I hope this helps explain, thanks! I am available to help.

**Comment by KrisNey-MSFT** [comment]
Hi @maheswari-s  - here is the comment from @chrispsommers above about trying out this DPDK version with the DASH build pipeline to make sure it compiles P4 with DPDK, including the 128-bit operations?

**Comment by jafingerhut** [comment]
This PR https://github.com/p4lang/p4c/pull/4952 that Sosutha mentioned earlier has already been merged into p4c.  I did not personally test that newer version compiling DASH reference model P4 code myself, before it was committed.  I do not know whether Sosutha tried that.

Best response I can give is: try it with latest version of p4c and see what happens, if you are interested.

**Comment by KrisNey-MSFT** [comment]
> This PR [p4lang/p4c#4952](https://github.com/p4lang/p4c/pull/4952) that Sosutha mentioned earlier has already been merged into p4c. I did not personally test that newer version compiling DASH reference model P4 code myself, before it was committed. I do not know whether Sosutha tried that.
> 
> Best response I can give is: try it with latest version of p4c and see what happens, if you are interested.

@jafingerhut - that would be awesome if you have the time - thank you!

**Comment by KrisNey-MSFT** [comment]
hi @jafingerhut - did you mean that you were going to try it?  Or someone/anyone could try it?
TY, K~

**Comment by jafingerhut** [comment]
Sorry, I was not planning on trying it.  My current guess is that unless someone else takes up more active development of the P4 DPDK back end and runtime API, that BMv2 is probably a more viable option, feature-wise.

**Comment by KrisNey-MSFT** [comment]
Got it, thanks Andy for clarifying.

**Comment by KrisNey-MSFT** [comment]
We are staying with bmv2 and will close this one.

**Comment by maheswari-s** [comment]
@chrispsommers and @KrisNey-MSFT,

I did the below changes in dash-pipeline folder:

```
diff --git a/dash-pipeline/Makefile b/dash-pipeline/Makefile
index 99bd1a9..73be4f2 100644
--- a/dash-pipeline/Makefile
+++ b/dash-pipeline/Makefile
-               -DTARGET_DPDK_PNA -DPNA_CONNTRACK -DDISABLE_128BIT_ARITHMETIC \
+               -DTARGET_DPDK_PNA -DPNA_CONNTRACK \

diff --git a/dash-pipeline/dockerfiles/DOCKER_P4C_DPDK_IMG.env b/dash-pipeline/dockerfiles/DOCKER_P4C_DPDK_IMG.env
index a4c8fdf..3b9c1f3 100644
--- a/dash-pipeline/dockerfiles/DOCKER_P4C_DPDK_IMG.env
+++ b/dash-pipeline/dockerfiles/DOCKER_P4C_DPDK_IMG.env
-export DOCKER_P4C_DPDK_IMG_CTAG?=240201
+export DOCKER_P4C_DPDK_IMG_CTAG?=250523

diff --git a/dash-pipeline/dockerfiles/Dockerfile.p4c-dpdk b/dash-pipeline/dockerfiles/Dockerfile.p4c-dpdk
index aa98bf2..11dc7c9 100644
--- a/dash-pipeline/dockerfiles/Dockerfile.p4c-dpdk
+++ b/dash-pipeline/dockerfiles/Dockerfile.p4c-dpdk
-# :latest on 2024-Feb-01:
-FROM p4lang/p4c:1.2.4.8 as p4lang-p4c
+# :latest on 2025-May-23:
+FROM p4lang/p4c@sha256:sha256:241b8878af0771a7bed896325807a51c4acdd4e3b7f3ed28ed1eb382890d60d6 as p4lang-p4c
+#FROM p4lang/p4c:1.2.4.8 as p4lang-p4c
```

I tried to do **make all run-switch** but failed at docker getting the ""https://sonicdash.azurecr.io/v2/". I guess there is some permission issue. 

Shall i just push this P4C SHA into the repo and someone can try compiling the DASH P4?
Otherwise, let me know how to get required permissions.

**Comment by chrispsommers** [comment]
@maheswari-s Can you show the actual error message please?

**Comment by maheswari-s** [comment]
> [@maheswari-s](https://github.com/maheswari-s) Can you show the actual error message please?

**First got this error:**
Unable to find image 'sonicdash.azurecr.io/dash-p4c-bmv2:40c82fd0c19' locally
docker: Error response from daemon: Get "https://sonicdash.azurecr.io/v2/": net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers).
See 'docker run --help'.

**I did docker stop and start. After that getting below error:**
Unable to find image 'sonicdash.azurecr.io/dash-p4c-bmv2:40c82fd0c19' locally
docker: Error response from daemon: Get "https://sonicdash.azurecr.io/v2/": context deadline exceeded (Client.Timeout exceeded while awaiting headers).
See 'docker run --help'.


**I copy pasted the URL in web browser, and got below message:**
{"errors":[{"code":"UNAUTHORIZED","message":"authentication required, visit https://aka.ms/acr/authorization for more information.","detail":null}]}


So i suspected that there is some permission issue.

**Comment by chrispsommers** [comment]
>Unable to find image 'sonicdash.azurecr.io/dash-p4c-bmv2:40c82fd0c19' locally
Are you building this image locally? You shouldn't need to fetch it.
Can you run this make target? [docker-dash-p4c-dpdk](https://github.com/sonic-net/DASH/blob/57e599c9d8e1f538ea01fad28781d8a0c2706ba5/dash-pipeline/Makefile#L632C1-L632C17)

You might want to add `ENABLE_DOCKER_PULL=` to your make command to ensure it doesn't try to pull it from the cloud. it's been a while since I did any of this.
